x-pack/elastic-agent/pkg/core/plugin/app/app.go

```diff
@@ -121,66 +131,61 @@ func (a *Application) Stop() {
 	a.appLock.Lock()
 	defer a.appLock.Unlock()
 
-	if a.state.Status == state.Running && a.state.ProcessInfo != nil {
-		if closeClient, ok := a.grpcClient.(closer); ok {
-			closeClient.Close()
-		}
-
-		process.Stop(a.logger, a.state.ProcessInfo.PID)
+	if a.state.Status == state.Stopped {
+		return
+	}
 
-		a.state.Status = state.Stopped
-		a.state.ProcessInfo = nil
-		a.grpcClient = nil
-
-		// remove generated configuration if present
-		filename := fmt.Sprintf(configFileTempl, a.id)
-		filePath, err := filepath.Abs(filepath.Join(a.downloadConfig.InstallPath, filename))
-		if err == nil {
-			// ignoring error: not critical
-			os.Remove(filePath)
+	stopSig := os.Interrupt
+	if a.srvState != nil {
+		if err := a.srvState.Stop(a.processConfig.StopTimeout); err != nil {
+			// kill the process if stop through GRPC doesn't work
+			stopSig = os.Kill
+		}
+		a.srvState = nil
+	}
+	if a.state.ProcessInfo != nil {
+		if err := a.state.ProcessInfo.Process.Signal(stopSig); err == nil {
+			// no error on signal, so wait for it to stop
+			_, _ = a.state.ProcessInfo.Process.Wait()
 		}
+		a.state.ProcessInfo = nil
 
 		// cleanup drops
 		a.monitor.Cleanup(a.name, a.pipelineID)
 	}
 }
 
-// State returns the state of the application [Running, Stopped].
-func (a *Application) State() state.State {
-	a.appLock.Lock()
-	defer a.appLock.Unlock()
-
-	return a.state
-}
-
-func (a *Application) watch(ctx context.Context, p Taggable, proc *os.Process, cfg map[string]interface{}) {
+func (a *Application) watch(ctx context.Context, p Taggable, proc *process.Info, cfg map[string]interface{}) {
 	go func() {
 		var procState *os.ProcessState
 
 		select {
-		case ps := <-a.waitProc(proc):
+		case ps := <-a.waitProc(proc.Process):
 			procState = ps
 		case <-a.bgContext.Done():
 			a.Stop()
 			return
 		}
 
 		a.appLock.Lock()
-		s := a.state.Status
-		a.state.Status = state.Stopped
 		a.state.ProcessInfo = nil
+		srvState := a.srvState
 		a.appLock.Unlock()
 
-		if procState.Success() {
+		if srvState.Expected() == proto.StateExpected_STOPPING {
+			a.state.Status = state.Stopped
```

Comment:
Could this become a concurrency issue? I see above on line 133 that we lock if we only read the state but here we also write it?

Reply:
Yes that `Unlock` should be moved down before the `a.Start`.